core/src/test/scala/org/apache/spark/sql/store/CreateIndexTest.scala

```diff
@@ -0,0 +1,772 @@
+      Map.empty[String, String])
+    snContext.dropIndex("test1", false)
+
+    snContext.createIndex("test1", tableName, Map(("col1" -> Some(Ascending))),
+      Map.empty[String, String])
+    snContext.dropIndex("test1", false)
+
+    // drop non-existent indexes with if exist clause
+    snContext.dropIndex("test1", true)
+    snContext.sql("drop index if exists test1")
+  }
+
+  private def createBase3Tables(snContext: SnappyContext,
+      table1: String,
+      table2: String,
+      table3: String): Unit = {
+
+    val props = Map(
+      "PARTITION_BY" -> "col1")
+    snContext.sql(s"drop table if exists $table1")
+
+    val data = Seq(Seq(111, "aaa", "hello"),
+      Seq(222, "bbb", "halo"),
+      Seq(333, "aaa", "hello"),
+      Seq(444, "bbb", "halo"),
+      Seq(555, "ccc", "halo"),
+      Seq(666, "ccc", "halo")
+    )
+
+    val rdd = sc.parallelize(data, data.length).map(s =>
+      new Data2(s(0).asInstanceOf[Int], s(1).asInstanceOf[String], s(2).asInstanceOf[String]))
+    val dataDF = snContext.createDataFrame(rdd)
+
+    snContext.createTable(s"$table2", "column", dataDF.schema,
+      props + ("PARTITION_BY" -> "col2,col3"))
+    tablesToDrop += table2
+
+    snContext.createTable(s"$table3", "column", dataDF.schema, props)
+    tablesToDrop += table3
+
+    dataDF.write.format("column").mode(SaveMode.Append).options(props).saveAsTable(table1)
+    tablesToDrop += table1
+
+    dataDF.write.insertInto(table2)
+    dataDF.write.insertInto(table3)
+  }
+
+  test("Test two table joins") {
+    val table1 = "tabOne"
+    val table2 = "tabTwo"
+    val table3 = "tabThree"
+
+    val index1 = s"${table1}_IdxOne"
+    val index2 = s"${table1}_IdxTwo"
+    val index3 = s"${table1}_IdxThree"
+    val index4 = s"${table1}_IdxFour"
+
+    val index31 = s"${table3}_IdxOne"
+
+    val leftIdx = Seq(1, 2)
+    val rightIdx = Seq(3, 4, 5, 6)
+
+    val snContext = SnappyContext(sc)
+    context.set(snContext)
+    snContext.setConf(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key, "-1")
+
+    createBase3Tables(snContext, table1, table2, table3)
+
+    doPrint("Verify index create and drop for various index types")
+    snContext.sql(s"create index $index1 on $table1 (COL1)")
+    indexesToDrop += index1
+    snContext.sql(s"create index $index2 on $table1 " +
+        s" (COL1) Options (colocate_with  '$table3')")
+    indexesToDrop += index2
+
+    snContext.sql(s"create index $index3 on $table1 (COL2, COL3)")
+    indexesToDrop += index3
+    snContext.sql(s"create index $index4 on $table1 " +
+        s" (COL2, COL3) Options (colocate_with  '$table2')")
+    indexesToDrop += index4
+
+    snContext.sql(s"create index $index31 on $table3 " +
+        s" (COL2, COL3) Options (colocate_with  '$index4')")
+    indexesToDrop += index31
+
+    val executeQ = QueryExecutor(snContext)
+
+    executeQ(s"select * from $table1 tab1 join $table3 tab2 on tab1.col1 = tab2.col1") {
+      validateIndex(Seq(index2), table3)(_)
```

Comment:
@ymahajan thanks for looking at the tests :) .. yes basis for index choice is most number of partitioning columns. Largest colocation chain overrides choice of index though.